feat: Reclassify nomcom feedback

pselkirk opened the pull request:
Fixes #4669

pselkirk committed:
fix: Clean up view_feedback_pending

- Remove "Unclassified" column header, which caused misalignment in the table body.

- Show the message author - previously displayed as `(None)`.

pselkirk committed:
feat: Reclassify nomcom feedback (#4669)

- There's a new `Chair/Advisor Tasks` menu item `Reclassify feedback`.

- I overloaded `view_feedback*` URLs with a `?reclassify` parameter.

- This adds a checkbox to each feedback message, and a `Reclassify` button
at the bottom of each feedback page.

- "Reclassifying" basically de-classifies the feedback, and punts it back
to the "Pending emails" view for reclassification.

- If a feedback has been applied to multiple nominees, declassifying it
from one nominee removes it from all.

pselkirk commented:
Still to be done:

- Test cases.
- Robert asked for an "overcome by events" classification.

pselkirk committed:
fix: Remove unused local variables

pselkirk committed:
fix: Fix some missing and mis-nested html

pselkirk committed:
test: Add tests for reclassifying feedback

richsalz commented:
So this adds a "reclassify" action, that puts already-classified feedback back in the work queue.  That's a reasonable approach. This looks fine to me.

jennifer-richards reviewed (changes requested):
I think the UI needs to indicate that it's changed from "viewing" to "reclassifying". As it is, it's hard to see that the "Reclassify feedback" menu item actually had any effect or that clicking on the individual items will do other than view the feedback.

I wonder if a button on each individual feedback item's view page would be easier to implement in a way that's clearly visible to the user. I.e., open the "View feedback" tab, click the feedback item, then click the "reclassify" button on it to go to a confirmation view.

If you stick with the batch reclassify mode, there are also some glitches on the `view_feedback_nominee.html` template - the feedback type tabs don't seem to do anything. That should be cleaned up.

pselkirk commented on ietf/nomcom/views.py:
```diff
@@ -956,15 +960,30 @@ def view_feedback_unrelated(request, year):
 @role_required("Nomcom")
 @nomcom_private_key_required
-def view_feedback_topic(request, year, topic_id):
+def view_feedback_topic(request, year, topic_id, reflassify=False):
```
Right, that's from an earlier iteration.

pselkirk commented on ietf/templates/nomcom/view_feedback_nominee.html:
```diff
@@ -20,11 +20,26 @@ <h2>Feedback about {{ nominee }}</h2>
                     {% if feedback.type.slug == ft.slug %}
                         <dl class="row">
+                            {% if is_chair and nomcom.group.state_id == 'active' and reclassify and nominee.feedback_set.all|length > 0 %}
+                            <dt class="col-sm-2">
```
An earlier iteration had the reclassification turned on full-time for the chair, and I felt that checkboxes would be less visually intrusive than a `Reclassify` button on each message. I eventually convinced myself that reclassification would be (or at least should be) infrequent enough that it should need to be explicitly enabled. Also for safety. So maybe now we want a `Reclassify` button at the bottom of each message, instead of a checkbox at the top.

jennifer-richards replied:
I later looked at a list that had more than one feedback item in the table and the checkbox was clearer than when I had only a single item, so maybe it's ok. Really the thing that concerned me was the lack of indication that the UI was in "reclassify mode". My suggestion of adding in the buttons could well add too much clutter - it was just a thought.

pselkirk committed:
refactor: Substantial redesign of feedback reclassification

- Break out reclassify_feedback* as their own URLs and views,
  and revert changes to view_feedback*.html.

- Replace checkboxes with a Reclassify button on each message.

pselkirk commented:
Robert asked for an "overcome by events" feedback type, which was one of the motivators for this feature, but I feel like it's orthogonal to the actual functionality of reclassification. That might be better handled in a separate ticket.

rjsparks commented:
It's what things will often be reclassified _to_.
I'd like to see it as part of this PR, but a companion that merged near the same time would be sufficient

pselkirk committed:
fix: Remember to clear the feedback associations when reclassifying

pselkirk committed:
feat: Add an 'Overcome by events' feedback type

pselkirk committed:
refactor: When invoking reclassification from a view-feedback page, load the corresponding reclassify-feedback page

pselkirk committed:
Merge branch 'main' into fix-4669

pselkirk committed:
fix: De-conflict migration with 0004_statements

Also change the coding style to match, and add a reverse migration.

pselkirk committed:
fix: Fix a test case to account for new feedback type

pselkirk committed:
fix: 842e730 broke the Back button

pselkirk committed:
refactor: Reclassify feedback directly instead of putting it back in the work queue

pselkirk committed:
fix: Adjust tests to new workflow

pselkirk committed:
refactor: Further refine reclassification to avoid redirects

pselkirk committed:
refactor: Impose a FeedbackTypeName ordering

Also add FeedbackTypeName.legend field, rather than synthesizing it every
time we classify or reclassify feedback.

In the reclassification forms, only show the relevant feedback types.

jennifer-richards commented on ietf/nomcom/views.py:
```diff
@@ -924,47 +939,85 @@ def view_feedback_pending(request, year):
+                messages.success(request, 'The selected feedback has been de-classified. Please reclassify it in the Pending emails tab.')
+            else:
+                feedback.type = FeedbackTypeName.objects.get(slug=type)
+                messages.success(request, 'The selected feedback has been reclassified.')
```
Minor, but maybe add the new classification type to the success message?

jennifer-richards commented on ietf/nomcom/views.py:
```diff
@@ -976,17 +1029,50 @@ def view_feedback_topic(request, year, topic_id):
+                messages.success(request, 'The selected feedback has been de-classified. Please reclassify it in the Pending emails tab.')
+            else:
+                feedback.type = FeedbackTypeName.objects.get(slug=type)
+                messages.success(request, 'The selected feedback has been reclassified.')
```
Again add the type to the success message?

jennifer-richards commented on ietf/templates/nomcom/reclassify_feedback_item.html:
```diff
@@ -0,0 +1,103 @@
+            </tbody>
+        </table>
+        <input type="hidden" name="feedback_id" value="{{ feedback.id }}">
+        <button class="btn btn-primary" type="submit">Classify</button>
```
I think the button should perhaps say "Save" rather than "Classify"?

pselkirk replied:
IDK, the form for pending email says "Classify", and this form is reached by the "Reclassify" button.

jennifer-richards replied:
Ok - I guess my issue is with the existing UI then. :-) Leave it matching.

jennifer-richards commented on ietf/nomcom/views.py:
```diff
@@ -924,47 +939,85 @@ def view_feedback_pending(request, year):
+        feedback.type = None
+        feedback.topics.clear()
+        feedback.save()
+        messages.success(request, 'The selected feedback has been de-classified. Please reclassify it in the Pending emails tab.')
```
Do you intentionally only handle clearing the feedback type for "topic" feedback? Why not allow reclassification as for the other flavors?

pselkirk replied:
The logic here is to reclassify a feedback item within its category (nominee, topic, unrelated). e.g. Nominee feedback can be reclassified from `questio` to `obe` (the canonical use case), while keeping its association with the nominee. Or it can be set to `unclassified`, in which case we wipe the type and nominees, and send it back to the pending queue.

Topic feedback only allows type `comment`, so the reclassification options would be `comment` (leave it unchanged) or `unclassified`. If the chair wants to reclassify a topic comment, the only reasonable option would seem to be to send it back to the pending queue, where it could be attached to a different topic, or marked as read, I don't know.

jennifer-richards reviewed (changes requested):
I have a number of small things, plus a couple non-trivial requests. One (the `HTTP_REFERER` business) I think needs to be sorted out now, the other (using a `ModelForm`) could wait if you're not enthusiastic to jump on it.

I like the "Reclassify" button format, but I'm wondering whether we really need the reclassify mode at all. Is there any reason not to, say, shrink the button a little and perhaps relocate it, then just always have it present? I think that would do away with the need to deal with the link in the base template and eliminate the need that led to the `HTTP_REFERER` business.

pselkirk commented:
> I like the "Reclassify" button format, but I'm wondering whether we really need the reclassify mode at all. Is there any reason not to, say, shrink the button a little and perhaps relocate it, then just always have it present?

My feeling was that reclassification would be uncommon enough that a) there should be an explicit action to initiate it, and b) having the button there for the 99.99% of the time that you're not reclassifying would be visual clutter.

OTOH, I'm not the target user. @richsalz @martinthomson

pselkirk commented on ietf/nomcom/views.py:
```diff
@@ -838,7 +851,9 @@ def nominee_staterank(nominee):
+                               'is_chair_task': reclassify,
+                               })
 
 
```
??

martinthomson commented:
I'd have to see this in action to be able to reach any sort of conclusion.

richsalz commented:
It's easy to ignore a button if it's always there.  If it comes and goes, that's a problem.

jennifer-richards commented:
> My feeling was that reclassification would be uncommon enough that a) there should be an explicit action to initiate it, and b) having the button there for the 99.99% of the time that you're not reclassifying would be visual clutter.

I agree it'll likely be uncommon, but I think that's a reason to keep it closer at hand in the usual view. With the button always present, when the chair comes across a piece of feedback that's incorrectly classified they won't have to remember to look through the "chair tasks" menu and then find the feedback again.

I think shrinking the button will reduce the clutter to a tolerable level. I can imagine we eventually associate a pulldown menu with each feedback item that has the reclassify view as one of a few options (but, emphatically, am not proposing that right now!).

As a bonus it'll simplify the views that are doing double-duty here.

pselkirk committed:
Merge branch 'main' into fix-4669

pselkirk committed:
refactor: Merge reclassify_feedback_* back into view_feedback_*

This means the "Reclassify" button is always present, but eliminates some
complexity.

pselkirk committed:
refactor: Add filter(used=True) on FeedbackTypeName querysets

pselkirk committed:
refactor: Add the new FeedbackTypeName to the reclassification success message

pselkirk committed:
fix: Secure reclassification against rogue nomcom members

jennifer-richards reviewed (approved):
Looks good to me!

Your worry about the "Reclassify" being a little obtrusive (even with the small button) is a reasonable one, but I think this is a good starting point. If it annoys the chair, we can restyle the view after we have some experience with how this is used.

richsalz reviewed (approved):
We don't need "button flavor of the month" :)  This is good.  Thanks!  And future nomcom chairs will thank you, too.
